// @mantleio/contracts/dist/types/factories/contracts/libraries/eigenda/interfaces/IBeaconChainOracle__factory.ts
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */

import { Contract, Signer, utils } from "ethers";
import type { Provider } from "@ethersproject/providers";
import type {
  IBeaconChainOracle,
  IBeaconChainOracleInterface,
} from "../../../../../contracts/libraries/eigenda/interfaces/IBeaconChainOracle";

const _abi = [
  {
    inputs: [],
    name: "getBeaconChainStateRoot",
    outputs: [
      {
        internalType: "bytes32",
        name: "",
        type: "bytes32",
      },
    ],
    stateMutability: "view",
    type: "function",
  },
  {
    inputs: [
      {
        internalType: "bytes32",
        name: "beaconChainStateRoot",
        type: "bytes32",
      },
    ],
    name: "setBeaconChainStateRoot",
    outputs: [],
    stateMutability: "nonpayable",
    type: "function",
  },
];

export class IBeaconChainOracle__factory {
  static readonly abi = _abi;
  static createInterface(): IBeaconChainOracleInterface {
    return new utils.Interface(_abi) as IBeaconChainOracleInterface;
  }
  static connect(
    address: string,
    signerOrProvider: Signer | Provider
  ): IBeaconChainOracle {
    return new Contract(address, _abi, signerOrProvider) as IBeaconChainOracle;
  }
}
