// @mantleio/contracts/dist/types/factories/contracts/libraries/eigenda/interfaces/IPauserRegistry__factory.ts
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */

import { Contract, Signer, utils } from "ethers";
import type { Provider } from "@ethersproject/providers";
import type {
  IPauserRegistry,
  IPauserRegistryInterface,
} from "../../../../../contracts/libraries/eigenda/interfaces/IPauserRegistry";

const _abi = [
  {
    inputs: [],
    name: "pauser",
    outputs: [
      {
        internalType: "address",
        name: "",
        type: "address",
      },
    ],
    stateMutability: "view",
    type: "function",
  },
  {
    inputs: [],
    name: "unpauser",
    outputs: [
      {
        internalType: "address",
        name: "",
        type: "address",
      },
    ],
    stateMutability: "view",
    type: "function",
  },
];

export class IPauserRegistry__factory {
  static readonly abi = _abi;
  static createInterface(): IPauserRegistryInterface {
    return new utils.Interface(_abi) as IPauserRegistryInterface;
  }
  static connect(
    address: string,
    signerOrProvider: Signer | Provider
  ): IPauserRegistry {
    return new Contract(address, _abi, signerOrProvider) as IPauserRegistry;
  }
}
