// @mantleio/contracts/dist/types/factories/contracts/libraries/eigenda/interfaces/ISlasher__factory.ts
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */

import { Contract, Signer, utils } from "ethers";
import type { Provider } from "@ethersproject/providers";
import type {
  ISlasher,
  ISlasherInterface,
} from "../../../../../contracts/libraries/eigenda/interfaces/ISlasher";

const _abi = [
  {
    inputs: [
      {
        internalType: "address",
        name: "operator",
        type: "address",
      },
      {
        internalType: "address",
        name: "serviceContract",
        type: "address",
      },
    ],
    name: "bondedUntil",
    outputs: [
      {
        internalType: "uint32",
        name: "",
        type: "uint32",
      },
    ],
    stateMutability: "view",
    type: "function",
  },
  {
    inputs: [
      {
        internalType: "address",
        name: "toBeSlashed",
        type: "address",
      },
      {
        internalType: "address",
        name: "slashingContract",
        type: "address",
      },
    ],
    name: "canSlash",
    outputs: [
      {
        internalType: "bool",
        name: "",
        type: "bool",
      },
    ],
    stateMutability: "view",
    type: "function",
  },
  {
    inputs: [
      {
        internalType: "address",
        name: "operator",
        type: "address",
      },
      {
        internalType: "uint32",
        name: "withdrawalStartBlock",
        type: "uint32",
      },
      {
        internalType: "uint256",
        name: "middlewareTimesIndex",
        type: "uint256",
      },
    ],
    name: "canWithdraw",
    outputs: [
      {
        internalType: "bool",
        name: "",
        type: "bool",
      },
    ],
    stateMutability: "nonpayable",
    type: "function",
  },
  {
    inputs: [
      {
        internalType: "address",
        name: "toBeFrozen",
        type: "address",
      },
    ],
    name: "freezeOperator",
    outputs: [],
    stateMutability: "nonpayable",
    type: "function",
  },
  {
    inputs: [
      {
        internalType: "address",
        name: "operator",
        type: "address",
      },
      {
        internalType: "uint32",
        name: "updateBlock",
        type: "uint32",
      },
    ],
    name: "getCorrectValueForInsertAfter",
    outputs: [
      {
        internalType: "uint256",
        name: "",
        type: "uint256",
      },
    ],
    stateMutability: "view",
    type: "function",
  },
  {
    inputs: [
      {
        internalType: "address",
        name: "operator",
        type: "address",
      },
      {
        internalType: "uint32",
        name: "index",
        type: "uint32",
      },
    ],
    name: "getMiddlewareTimesIndexBlock",
    outputs: [
      {
        internalType: "uint32",
        name: "",
        type: "uint32",
      },
    ],
    stateMutability: "view",
    type: "function",
  },
  {
    inputs: [
      {
        internalType: "address",
        name: "operator",
        type: "address",
      },
      {
        internalType: "uint32",
        name: "index",
        type: "uint32",
      },
    ],
    name: "getMiddlewareTimesIndexServeUntil",
    outputs: [
      {
        internalType: "uint32",
        name: "",
        type: "uint32",
      },
    ],
    stateMutability: "view",
    type: "function",
  },
  {
    inputs: [
      {
        internalType: "address",
        name: "staker",
        type: "address",
      },
    ],
    name: "isFrozen",
    outputs: [
      {
        internalType: "bool",
        name: "",
        type: "bool",
      },
    ],
    stateMutability: "view",
    type: "function",
  },
  {
    inputs: [
      {
        internalType: "address",
        name: "operator",
        type: "address",
      },
      {
        internalType: "address",
        name: "serviceContract",
        type: "address",
      },
    ],
    name: "latestUpdateBlock",
    outputs: [
      {
        internalType: "uint32",
        name: "",
        type: "uint32",
      },
    ],
    stateMutability: "view",
    type: "function",
  },
  {
    inputs: [
      {
        internalType: "address",
        name: "contractAddress",
        type: "address",
      },
    ],
    name: "optIntoSlashing",
    outputs: [],
    stateMutability: "nonpayable",
    type: "function",
  },
  {
    inputs: [
      {
        internalType: "address",
        name: "operator",
        type: "address",
      },
      {
        internalType: "uint32",
        name: "serveUntil",
        type: "uint32",
      },
    ],
    name: "recordFirstStakeUpdate",
    outputs: [],
    stateMutability: "nonpayable",
    type: "function",
  },
  {
    inputs: [
      {
        internalType: "address",
        name: "operator",
        type: "address",
      },
      {
        internalType: "uint32",
        name: "serveUntil",
        type: "uint32",
      },
    ],
    name: "recordLastStakeUpdateAndRevokeSlashingAbility",
    outputs: [],
    stateMutability: "nonpayable",
    type: "function",
  },
  {
    inputs: [
      {
        internalType: "address",
        name: "operator",
        type: "address",
      },
      {
        internalType: "uint32",
        name: "updateBlock",
        type: "uint32",
      },
      {
        internalType: "uint32",
        name: "serveUntil",
        type: "uint32",
      },
      {
        internalType: "uint256",
        name: "insertAfter",
        type: "uint256",
      },
    ],
    name: "recordStakeUpdate",
    outputs: [],
    stateMutability: "nonpayable",
    type: "function",
  },
  {
    inputs: [
      {
        internalType: "address[]",
        name: "frozenAddresses",
        type: "address[]",
      },
    ],
    name: "resetFrozenStatus",
    outputs: [],
    stateMutability: "nonpayable",
    type: "function",
  },
];

export class ISlasher__factory {
  static readonly abi = _abi;
  static createInterface(): ISlasherInterface {
    return new utils.Interface(_abi) as ISlasherInterface;
  }
  static connect(
    address: string,
    signerOrProvider: Signer | Provider
  ): ISlasher {
    return new Contract(address, _abi, signerOrProvider) as ISlasher;
  }
}
