// @mantleio/contracts/dist/types/contracts/libraries/eigenda/interfaces/IBeaconChainOracle.ts
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
import type {
  BaseContract,
  BigNumber,
  BytesLike,
  CallOverrides,
  ContractTransaction,
  Overrides,
  PopulatedTransaction,
  Signer,
  utils,
} from "ethers";
import type { FunctionFragment, Result } from "@ethersproject/abi";
import type { Listener, Provider } from "@ethersproject/providers";
import type {
  TypedEventFilter,
  TypedEvent,
  TypedListener,
  OnEvent,
  PromiseOrValue,
} from "../../../../common";

export interface IBeaconChainOracleInterface extends utils.Interface {
  functions: {
    "getBeaconChainStateRoot()": FunctionFragment;
    "setBeaconChainStateRoot(bytes32)": FunctionFragment;
  };

  getFunction(
    nameOrSignatureOrTopic:
      | "getBeaconChainStateRoot"
      | "setBeaconChainStateRoot"
  ): FunctionFragment;

  encodeFunctionData(
    functionFragment: "getBeaconChainStateRoot",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "setBeaconChainStateRoot",
    values: [PromiseOrValue<BytesLike>]
  ): string;

  decodeFunctionResult(
    functionFragment: "getBeaconChainStateRoot",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "setBeaconChainStateRoot",
    data: BytesLike
  ): Result;

  events: {};
}

export interface IBeaconChainOracle extends BaseContract {
  connect(signerOrProvider: Signer | Provider | string): this;
  attach(addressOrName: string): this;
  deployed(): Promise<this>;

  interface: IBeaconChainOracleInterface;

  queryFilter<TEvent extends TypedEvent>(
    event: TypedEventFilter<TEvent>,
    fromBlockOrBlockhash?: string | number | undefined,
    toBlock?: string | number | undefined
  ): Promise<Array<TEvent>>;

  listeners<TEvent extends TypedEvent>(
    eventFilter?: TypedEventFilter<TEvent>
  ): Array<TypedListener<TEvent>>;
  listeners(eventName?: string): Array<Listener>;
  removeAllListeners<TEvent extends TypedEvent>(
    eventFilter: TypedEventFilter<TEvent>
  ): this;
  removeAllListeners(eventName?: string): this;
  off: OnEvent<this>;
  on: OnEvent<this>;
  once: OnEvent<this>;
  removeListener: OnEvent<this>;

  functions: {
    getBeaconChainStateRoot(overrides?: CallOverrides): Promise<[string]>;

    setBeaconChainStateRoot(
      beaconChainStateRoot: PromiseOrValue<BytesLike>,
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<ContractTransaction>;
  };

  getBeaconChainStateRoot(overrides?: CallOverrides): Promise<string>;

  setBeaconChainStateRoot(
    beaconChainStateRoot: PromiseOrValue<BytesLike>,
    overrides?: Overrides & { from?: PromiseOrValue<string> }
  ): Promise<ContractTransaction>;

  callStatic: {
    getBeaconChainStateRoot(overrides?: CallOverrides): Promise<string>;

    setBeaconChainStateRoot(
      beaconChainStateRoot: PromiseOrValue<BytesLike>,
      overrides?: CallOverrides
    ): Promise<void>;
  };

  filters: {};

  estimateGas: {
    getBeaconChainStateRoot(overrides?: CallOverrides): Promise<BigNumber>;

    setBeaconChainStateRoot(
      beaconChainStateRoot: PromiseOrValue<BytesLike>,
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<BigNumber>;
  };

  populateTransaction: {
    getBeaconChainStateRoot(
      overrides?: CallOverrides
    ): Promise<PopulatedTransaction>;

    setBeaconChainStateRoot(
      beaconChainStateRoot: PromiseOrValue<BytesLike>,
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<PopulatedTransaction>;
  };
}
