// @mantleio/contracts/dist/types/contracts/libraries/eigenda/interfaces/IInvestmentManager.ts
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
import type {
  BaseContract,
  BigNumber,
  BigNumberish,
  BytesLike,
  CallOverrides,
  ContractTransaction,
  Overrides,
  PopulatedTransaction,
  Signer,
  utils,
} from "ethers";
import type { FunctionFragment, Result } from "@ethersproject/abi";
import type { Listener, Provider } from "@ethersproject/providers";
import type {
  TypedEventFilter,
  TypedEvent,
  TypedListener,
  OnEvent,
  PromiseOrValue,
} from "../../../../common";

export declare namespace IInvestmentManager {
  export type WithdrawerAndNonceStruct = {
    withdrawer: PromiseOrValue<string>;
    nonce: PromiseOrValue<BigNumberish>;
  };

  export type WithdrawerAndNonceStructOutput = [string, BigNumber] & {
    withdrawer: string;
    nonce: BigNumber;
  };

  export type QueuedWithdrawalStruct = {
    strategies: PromiseOrValue<string>[];
    tokens: PromiseOrValue<string>[];
    shares: PromiseOrValue<BigNumberish>[];
    depositor: PromiseOrValue<string>;
    withdrawerAndNonce: IInvestmentManager.WithdrawerAndNonceStruct;
    withdrawalStartBlock: PromiseOrValue<BigNumberish>;
    delegatedAddress: PromiseOrValue<string>;
  };

  export type QueuedWithdrawalStructOutput = [
    string[],
    string[],
    BigNumber[],
    string,
    IInvestmentManager.WithdrawerAndNonceStructOutput,
    number,
    string
  ] & {
    strategies: string[];
    tokens: string[];
    shares: BigNumber[];
    depositor: string;
    withdrawerAndNonce: IInvestmentManager.WithdrawerAndNonceStructOutput;
    withdrawalStartBlock: number;
    delegatedAddress: string;
  };
}

export interface IInvestmentManagerInterface extends utils.Interface {
  functions: {
    "beaconChainETHStrategy()": FunctionFragment;
    "calculateWithdrawalRoot((address[],address[],uint256[],address,(address,uint96),uint32,address))": FunctionFragment;
    "completeQueuedWithdrawal((address[],address[],uint256[],address,(address,uint96),uint32,address),uint256,bool)": FunctionFragment;
    "delegation()": FunctionFragment;
    "depositBeaconChainETH(address,uint256)": FunctionFragment;
    "depositIntoStrategy(address,address,uint256)": FunctionFragment;
    "depositIntoStrategyOnBehalfOf(address,address,uint256,address,uint256,bytes32,bytes32)": FunctionFragment;
    "getDeposits(address)": FunctionFragment;
    "investorStratShares(address,address)": FunctionFragment;
    "investorStratsLength(address)": FunctionFragment;
    "queueWithdrawal(uint256[],address[],address[],uint256[],address,bool)": FunctionFragment;
    "recordOvercommittedBeaconChainETH(address,uint256,uint256)": FunctionFragment;
    "slashQueuedWithdrawal(address,(address[],address[],uint256[],address,(address,uint96),uint32,address))": FunctionFragment;
    "slashShares(address,address,address[],address[],uint256[],uint256[])": FunctionFragment;
    "slasher()": FunctionFragment;
  };

  getFunction(
    nameOrSignatureOrTopic:
      | "beaconChainETHStrategy"
      | "calculateWithdrawalRoot"
      | "completeQueuedWithdrawal"
      | "delegation"
      | "depositBeaconChainETH"
      | "depositIntoStrategy"
      | "depositIntoStrategyOnBehalfOf"
      | "getDeposits"
      | "investorStratShares"
      | "investorStratsLength"
      | "queueWithdrawal"
      | "recordOvercommittedBeaconChainETH"
      | "slashQueuedWithdrawal"
      | "slashShares"
      | "slasher"
  ): FunctionFragment;

  encodeFunctionData(
    functionFragment: "beaconChainETHStrategy",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "calculateWithdrawalRoot",
    values: [IInvestmentManager.QueuedWithdrawalStruct]
  ): string;
  encodeFunctionData(
    functionFragment: "completeQueuedWithdrawal",
    values: [
      IInvestmentManager.QueuedWithdrawalStruct,
      PromiseOrValue<BigNumberish>,
      PromiseOrValue<boolean>
    ]
  ): string;
  encodeFunctionData(
    functionFragment: "delegation",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "depositBeaconChainETH",
    values: [PromiseOrValue<string>, PromiseOrValue<BigNumberish>]
  ): string;
  encodeFunctionData(
    functionFragment: "depositIntoStrategy",
    values: [
      PromiseOrValue<string>,
      PromiseOrValue<string>,
      PromiseOrValue<BigNumberish>
    ]
  ): string;
  encodeFunctionData(
    functionFragment: "depositIntoStrategyOnBehalfOf",
    values: [
      PromiseOrValue<string>,
      PromiseOrValue<string>,
      PromiseOrValue<BigNumberish>,
      PromiseOrValue<string>,
      PromiseOrValue<BigNumberish>,
      PromiseOrValue<BytesLike>,
      PromiseOrValue<BytesLike>
    ]
  ): string;
  encodeFunctionData(
    functionFragment: "getDeposits",
    values: [PromiseOrValue<string>]
  ): string;
  encodeFunctionData(
    functionFragment: "investorStratShares",
    values: [PromiseOrValue<string>, PromiseOrValue<string>]
  ): string;
  encodeFunctionData(
    functionFragment: "investorStratsLength",
    values: [PromiseOrValue<string>]
  ): string;
  encodeFunctionData(
    functionFragment: "queueWithdrawal",
    values: [
      PromiseOrValue<BigNumberish>[],
      PromiseOrValue<string>[],
      PromiseOrValue<string>[],
      PromiseOrValue<BigNumberish>[],
      PromiseOrValue<string>,
      PromiseOrValue<boolean>
    ]
  ): string;
  encodeFunctionData(
    functionFragment: "recordOvercommittedBeaconChainETH",
    values: [
      PromiseOrValue<string>,
      PromiseOrValue<BigNumberish>,
      PromiseOrValue<BigNumberish>
    ]
  ): string;
  encodeFunctionData(
    functionFragment: "slashQueuedWithdrawal",
    values: [PromiseOrValue<string>, IInvestmentManager.QueuedWithdrawalStruct]
  ): string;
  encodeFunctionData(
    functionFragment: "slashShares",
    values: [
      PromiseOrValue<string>,
      PromiseOrValue<string>,
      PromiseOrValue<string>[],
      PromiseOrValue<string>[],
      PromiseOrValue<BigNumberish>[],
      PromiseOrValue<BigNumberish>[]
    ]
  ): string;
  encodeFunctionData(functionFragment: "slasher", values?: undefined): string;

  decodeFunctionResult(
    functionFragment: "beaconChainETHStrategy",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "calculateWithdrawalRoot",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "completeQueuedWithdrawal",
    data: BytesLike
  ): Result;
  decodeFunctionResult(functionFragment: "delegation", data: BytesLike): Result;
  decodeFunctionResult(
    functionFragment: "depositBeaconChainETH",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "depositIntoStrategy",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "depositIntoStrategyOnBehalfOf",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "getDeposits",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "investorStratShares",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "investorStratsLength",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "queueWithdrawal",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "recordOvercommittedBeaconChainETH",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "slashQueuedWithdrawal",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "slashShares",
    data: BytesLike
  ): Result;
  decodeFunctionResult(functionFragment: "slasher", data: BytesLike): Result;

  events: {};
}

export interface IInvestmentManager extends BaseContract {
  connect(signerOrProvider: Signer | Provider | string): this;
  attach(addressOrName: string): this;
  deployed(): Promise<this>;

  interface: IInvestmentManagerInterface;

  queryFilter<TEvent extends TypedEvent>(
    event: TypedEventFilter<TEvent>,
    fromBlockOrBlockhash?: string | number | undefined,
    toBlock?: string | number | undefined
  ): Promise<Array<TEvent>>;

  listeners<TEvent extends TypedEvent>(
    eventFilter?: TypedEventFilter<TEvent>
  ): Array<TypedListener<TEvent>>;
  listeners(eventName?: string): Array<Listener>;
  removeAllListeners<TEvent extends TypedEvent>(
    eventFilter: TypedEventFilter<TEvent>
  ): this;
  removeAllListeners(eventName?: string): this;
  off: OnEvent<this>;
  on: OnEvent<this>;
  once: OnEvent<this>;
  removeListener: OnEvent<this>;

  functions: {
    beaconChainETHStrategy(overrides?: CallOverrides): Promise<[string]>;

    calculateWithdrawalRoot(
      queuedWithdrawal: IInvestmentManager.QueuedWithdrawalStruct,
      overrides?: CallOverrides
    ): Promise<[string]>;

    completeQueuedWithdrawal(
      queuedWithdrawal: IInvestmentManager.QueuedWithdrawalStruct,
      middlewareTimesIndex: PromiseOrValue<BigNumberish>,
      receiveAsTokens: PromiseOrValue<boolean>,
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<ContractTransaction>;

    delegation(overrides?: CallOverrides): Promise<[string]>;

    depositBeaconChainETH(
      staker: PromiseOrValue<string>,
      amount: PromiseOrValue<BigNumberish>,
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<ContractTransaction>;

    depositIntoStrategy(
      strategy: PromiseOrValue<string>,
      token: PromiseOrValue<string>,
      amount: PromiseOrValue<BigNumberish>,
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<ContractTransaction>;

    depositIntoStrategyOnBehalfOf(
      strategy: PromiseOrValue<string>,
      token: PromiseOrValue<string>,
      amount: PromiseOrValue<BigNumberish>,
      staker: PromiseOrValue<string>,
      expiry: PromiseOrValue<BigNumberish>,
      r: PromiseOrValue<BytesLike>,
      vs: PromiseOrValue<BytesLike>,
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<ContractTransaction>;

    getDeposits(
      depositor: PromiseOrValue<string>,
      overrides?: CallOverrides
    ): Promise<[string[], BigNumber[]]>;

    investorStratShares(
      user: PromiseOrValue<string>,
      strategy: PromiseOrValue<string>,
      overrides?: CallOverrides
    ): Promise<[BigNumber] & { shares: BigNumber }>;

    investorStratsLength(
      staker: PromiseOrValue<string>,
      overrides?: CallOverrides
    ): Promise<[BigNumber]>;

    queueWithdrawal(
      strategyIndexes: PromiseOrValue<BigNumberish>[],
      strategies: PromiseOrValue<string>[],
      tokens: PromiseOrValue<string>[],
      shareAmounts: PromiseOrValue<BigNumberish>[],
      withdrawer: PromiseOrValue<string>,
      undelegateIfPossible: PromiseOrValue<boolean>,
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<ContractTransaction>;

    recordOvercommittedBeaconChainETH(
      overcommittedPodOwner: PromiseOrValue<string>,
      beaconChainETHStrategyIndex: PromiseOrValue<BigNumberish>,
      amount: PromiseOrValue<BigNumberish>,
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<ContractTransaction>;

    slashQueuedWithdrawal(
      recipient: PromiseOrValue<string>,
      queuedWithdrawal: IInvestmentManager.QueuedWithdrawalStruct,
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<ContractTransaction>;

    slashShares(
      slashedAddress: PromiseOrValue<string>,
      recipient: PromiseOrValue<string>,
      strategies: PromiseOrValue<string>[],
      tokens: PromiseOrValue<string>[],
      strategyIndexes: PromiseOrValue<BigNumberish>[],
      shareAmounts: PromiseOrValue<BigNumberish>[],
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<ContractTransaction>;

    slasher(overrides?: CallOverrides): Promise<[string]>;
  };

  beaconChainETHStrategy(overrides?: CallOverrides): Promise<string>;

  calculateWithdrawalRoot(
    queuedWithdrawal: IInvestmentManager.QueuedWithdrawalStruct,
    overrides?: CallOverrides
  ): Promise<string>;

  completeQueuedWithdrawal(
    queuedWithdrawal: IInvestmentManager.QueuedWithdrawalStruct,
    middlewareTimesIndex: PromiseOrValue<BigNumberish>,
    receiveAsTokens: PromiseOrValue<boolean>,
    overrides?: Overrides & { from?: PromiseOrValue<string> }
  ): Promise<ContractTransaction>;

  delegation(overrides?: CallOverrides): Promise<string>;

  depositBeaconChainETH(
    staker: PromiseOrValue<string>,
    amount: PromiseOrValue<BigNumberish>,
    overrides?: Overrides & { from?: PromiseOrValue<string> }
  ): Promise<ContractTransaction>;

  depositIntoStrategy(
    strategy: PromiseOrValue<string>,
    token: PromiseOrValue<string>,
    amount: PromiseOrValue<BigNumberish>,
    overrides?: Overrides & { from?: PromiseOrValue<string> }
  ): Promise<ContractTransaction>;

  depositIntoStrategyOnBehalfOf(
    strategy: PromiseOrValue<string>,
    token: PromiseOrValue<string>,
    amount: PromiseOrValue<BigNumberish>,
    staker: PromiseOrValue<string>,
    expiry: PromiseOrValue<BigNumberish>,
    r: PromiseOrValue<BytesLike>,
    vs: PromiseOrValue<BytesLike>,
    overrides?: Overrides & { from?: PromiseOrValue<string> }
  ): Promise<ContractTransaction>;

  getDeposits(
    depositor: PromiseOrValue<string>,
    overrides?: CallOverrides
  ): Promise<[string[], BigNumber[]]>;

  investorStratShares(
    user: PromiseOrValue<string>,
    strategy: PromiseOrValue<string>,
    overrides?: CallOverrides
  ): Promise<BigNumber>;

  investorStratsLength(
    staker: PromiseOrValue<string>,
    overrides?: CallOverrides
  ): Promise<BigNumber>;

  queueWithdrawal(
    strategyIndexes: PromiseOrValue<BigNumberish>[],
    strategies: PromiseOrValue<string>[],
    tokens: PromiseOrValue<string>[],
    shareAmounts: PromiseOrValue<BigNumberish>[],
    withdrawer: PromiseOrValue<string>,
    undelegateIfPossible: PromiseOrValue<boolean>,
    overrides?: Overrides & { from?: PromiseOrValue<string> }
  ): Promise<ContractTransaction>;

  recordOvercommittedBeaconChainETH(
    overcommittedPodOwner: PromiseOrValue<string>,
    beaconChainETHStrategyIndex: PromiseOrValue<BigNumberish>,
    amount: PromiseOrValue<BigNumberish>,
    overrides?: Overrides & { from?: PromiseOrValue<string> }
  ): Promise<ContractTransaction>;

  slashQueuedWithdrawal(
    recipient: PromiseOrValue<string>,
    queuedWithdrawal: IInvestmentManager.QueuedWithdrawalStruct,
    overrides?: Overrides & { from?: PromiseOrValue<string> }
  ): Promise<ContractTransaction>;

  slashShares(
    slashedAddress: PromiseOrValue<string>,
    recipient: PromiseOrValue<string>,
    strategies: PromiseOrValue<string>[],
    tokens: PromiseOrValue<string>[],
    strategyIndexes: PromiseOrValue<BigNumberish>[],
    shareAmounts: PromiseOrValue<BigNumberish>[],
    overrides?: Overrides & { from?: PromiseOrValue<string> }
  ): Promise<ContractTransaction>;

  slasher(overrides?: CallOverrides): Promise<string>;

  callStatic: {
    beaconChainETHStrategy(overrides?: CallOverrides): Promise<string>;

    calculateWithdrawalRoot(
      queuedWithdrawal: IInvestmentManager.QueuedWithdrawalStruct,
      overrides?: CallOverrides
    ): Promise<string>;

    completeQueuedWithdrawal(
      queuedWithdrawal: IInvestmentManager.QueuedWithdrawalStruct,
      middlewareTimesIndex: PromiseOrValue<BigNumberish>,
      receiveAsTokens: PromiseOrValue<boolean>,
      overrides?: CallOverrides
    ): Promise<void>;

    delegation(overrides?: CallOverrides): Promise<string>;

    depositBeaconChainETH(
      staker: PromiseOrValue<string>,
      amount: PromiseOrValue<BigNumberish>,
      overrides?: CallOverrides
    ): Promise<void>;

    depositIntoStrategy(
      strategy: PromiseOrValue<string>,
      token: PromiseOrValue<string>,
      amount: PromiseOrValue<BigNumberish>,
      overrides?: CallOverrides
    ): Promise<BigNumber>;

    depositIntoStrategyOnBehalfOf(
      strategy: PromiseOrValue<string>,
      token: PromiseOrValue<string>,
      amount: PromiseOrValue<BigNumberish>,
      staker: PromiseOrValue<string>,
      expiry: PromiseOrValue<BigNumberish>,
      r: PromiseOrValue<BytesLike>,
      vs: PromiseOrValue<BytesLike>,
      overrides?: CallOverrides
    ): Promise<BigNumber>;

    getDeposits(
      depositor: PromiseOrValue<string>,
      overrides?: CallOverrides
    ): Promise<[string[], BigNumber[]]>;

    investorStratShares(
      user: PromiseOrValue<string>,
      strategy: PromiseOrValue<string>,
      overrides?: CallOverrides
    ): Promise<BigNumber>;

    investorStratsLength(
      staker: PromiseOrValue<string>,
      overrides?: CallOverrides
    ): Promise<BigNumber>;

    queueWithdrawal(
      strategyIndexes: PromiseOrValue<BigNumberish>[],
      strategies: PromiseOrValue<string>[],
      tokens: PromiseOrValue<string>[],
      shareAmounts: PromiseOrValue<BigNumberish>[],
      withdrawer: PromiseOrValue<string>,
      undelegateIfPossible: PromiseOrValue<boolean>,
      overrides?: CallOverrides
    ): Promise<string>;

    recordOvercommittedBeaconChainETH(
      overcommittedPodOwner: PromiseOrValue<string>,
      beaconChainETHStrategyIndex: PromiseOrValue<BigNumberish>,
      amount: PromiseOrValue<BigNumberish>,
      overrides?: CallOverrides
    ): Promise<void>;

    slashQueuedWithdrawal(
      recipient: PromiseOrValue<string>,
      queuedWithdrawal: IInvestmentManager.QueuedWithdrawalStruct,
      overrides?: CallOverrides
    ): Promise<void>;

    slashShares(
      slashedAddress: PromiseOrValue<string>,
      recipient: PromiseOrValue<string>,
      strategies: PromiseOrValue<string>[],
      tokens: PromiseOrValue<string>[],
      strategyIndexes: PromiseOrValue<BigNumberish>[],
      shareAmounts: PromiseOrValue<BigNumberish>[],
      overrides?: CallOverrides
    ): Promise<void>;

    slasher(overrides?: CallOverrides): Promise<string>;
  };

  filters: {};

  estimateGas: {
    beaconChainETHStrategy(overrides?: CallOverrides): Promise<BigNumber>;

    calculateWithdrawalRoot(
      queuedWithdrawal: IInvestmentManager.QueuedWithdrawalStruct,
      overrides?: CallOverrides
    ): Promise<BigNumber>;

    completeQueuedWithdrawal(
      queuedWithdrawal: IInvestmentManager.QueuedWithdrawalStruct,
      middlewareTimesIndex: PromiseOrValue<BigNumberish>,
      receiveAsTokens: PromiseOrValue<boolean>,
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<BigNumber>;

    delegation(overrides?: CallOverrides): Promise<BigNumber>;

    depositBeaconChainETH(
      staker: PromiseOrValue<string>,
      amount: PromiseOrValue<BigNumberish>,
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<BigNumber>;

    depositIntoStrategy(
      strategy: PromiseOrValue<string>,
      token: PromiseOrValue<string>,
      amount: PromiseOrValue<BigNumberish>,
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<BigNumber>;

    depositIntoStrategyOnBehalfOf(
      strategy: PromiseOrValue<string>,
      token: PromiseOrValue<string>,
      amount: PromiseOrValue<BigNumberish>,
      staker: PromiseOrValue<string>,
      expiry: PromiseOrValue<BigNumberish>,
      r: PromiseOrValue<BytesLike>,
      vs: PromiseOrValue<BytesLike>,
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<BigNumber>;

    getDeposits(
      depositor: PromiseOrValue<string>,
      overrides?: CallOverrides
    ): Promise<BigNumber>;

    investorStratShares(
      user: PromiseOrValue<string>,
      strategy: PromiseOrValue<string>,
      overrides?: CallOverrides
    ): Promise<BigNumber>;

    investorStratsLength(
      staker: PromiseOrValue<string>,
      overrides?: CallOverrides
    ): Promise<BigNumber>;

    queueWithdrawal(
      strategyIndexes: PromiseOrValue<BigNumberish>[],
      strategies: PromiseOrValue<string>[],
      tokens: PromiseOrValue<string>[],
      shareAmounts: PromiseOrValue<BigNumberish>[],
      withdrawer: PromiseOrValue<string>,
      undelegateIfPossible: PromiseOrValue<boolean>,
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<BigNumber>;

    recordOvercommittedBeaconChainETH(
      overcommittedPodOwner: PromiseOrValue<string>,
      beaconChainETHStrategyIndex: PromiseOrValue<BigNumberish>,
      amount: PromiseOrValue<BigNumberish>,
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<BigNumber>;

    slashQueuedWithdrawal(
      recipient: PromiseOrValue<string>,
      queuedWithdrawal: IInvestmentManager.QueuedWithdrawalStruct,
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<BigNumber>;

    slashShares(
      slashedAddress: PromiseOrValue<string>,
      recipient: PromiseOrValue<string>,
      strategies: PromiseOrValue<string>[],
      tokens: PromiseOrValue<string>[],
      strategyIndexes: PromiseOrValue<BigNumberish>[],
      shareAmounts: PromiseOrValue<BigNumberish>[],
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<BigNumber>;

    slasher(overrides?: CallOverrides): Promise<BigNumber>;
  };

  populateTransaction: {
    beaconChainETHStrategy(
      overrides?: CallOverrides
    ): Promise<PopulatedTransaction>;

    calculateWithdrawalRoot(
      queuedWithdrawal: IInvestmentManager.QueuedWithdrawalStruct,
      overrides?: CallOverrides
    ): Promise<PopulatedTransaction>;

    completeQueuedWithdrawal(
      queuedWithdrawal: IInvestmentManager.QueuedWithdrawalStruct,
      middlewareTimesIndex: PromiseOrValue<BigNumberish>,
      receiveAsTokens: PromiseOrValue<boolean>,
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<PopulatedTransaction>;

    delegation(overrides?: CallOverrides): Promise<PopulatedTransaction>;

    depositBeaconChainETH(
      staker: PromiseOrValue<string>,
      amount: PromiseOrValue<BigNumberish>,
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<PopulatedTransaction>;

    depositIntoStrategy(
      strategy: PromiseOrValue<string>,
      token: PromiseOrValue<string>,
      amount: PromiseOrValue<BigNumberish>,
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<PopulatedTransaction>;

    depositIntoStrategyOnBehalfOf(
      strategy: PromiseOrValue<string>,
      token: PromiseOrValue<string>,
      amount: PromiseOrValue<BigNumberish>,
      staker: PromiseOrValue<string>,
      expiry: PromiseOrValue<BigNumberish>,
      r: PromiseOrValue<BytesLike>,
      vs: PromiseOrValue<BytesLike>,
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<PopulatedTransaction>;

    getDeposits(
      depositor: PromiseOrValue<string>,
      overrides?: CallOverrides
    ): Promise<PopulatedTransaction>;

    investorStratShares(
      user: PromiseOrValue<string>,
      strategy: PromiseOrValue<string>,
      overrides?: CallOverrides
    ): Promise<PopulatedTransaction>;

    investorStratsLength(
      staker: PromiseOrValue<string>,
      overrides?: CallOverrides
    ): Promise<PopulatedTransaction>;

    queueWithdrawal(
      strategyIndexes: PromiseOrValue<BigNumberish>[],
      strategies: PromiseOrValue<string>[],
      tokens: PromiseOrValue<string>[],
      shareAmounts: PromiseOrValue<BigNumberish>[],
      withdrawer: PromiseOrValue<string>,
      undelegateIfPossible: PromiseOrValue<boolean>,
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<PopulatedTransaction>;

    recordOvercommittedBeaconChainETH(
      overcommittedPodOwner: PromiseOrValue<string>,
      beaconChainETHStrategyIndex: PromiseOrValue<BigNumberish>,
      amount: PromiseOrValue<BigNumberish>,
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<PopulatedTransaction>;

    slashQueuedWithdrawal(
      recipient: PromiseOrValue<string>,
      queuedWithdrawal: IInvestmentManager.QueuedWithdrawalStruct,
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<PopulatedTransaction>;

    slashShares(
      slashedAddress: PromiseOrValue<string>,
      recipient: PromiseOrValue<string>,
      strategies: PromiseOrValue<string>[],
      tokens: PromiseOrValue<string>[],
      strategyIndexes: PromiseOrValue<BigNumberish>[],
      shareAmounts: PromiseOrValue<BigNumberish>[],
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<PopulatedTransaction>;

    slasher(overrides?: CallOverrides): Promise<PopulatedTransaction>;
  };
}
