// @mantleio/contracts/dist/types/contracts/libraries/eigenda/interfaces/IPauserRegistry.ts
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
import type {
  BaseContract,
  BigNumber,
  BytesLike,
  CallOverrides,
  PopulatedTransaction,
  Signer,
  utils,
} from "ethers";
import type { FunctionFragment, Result } from "@ethersproject/abi";
import type { Listener, Provider } from "@ethersproject/providers";
import type {
  TypedEventFilter,
  TypedEvent,
  TypedListener,
  OnEvent,
  PromiseOrValue,
} from "../../../../common";

export interface IPauserRegistryInterface extends utils.Interface {
  functions: {
    "pauser()": FunctionFragment;
    "unpauser()": FunctionFragment;
  };

  getFunction(nameOrSignatureOrTopic: "pauser" | "unpauser"): FunctionFragment;

  encodeFunctionData(functionFragment: "pauser", values?: undefined): string;
  encodeFunctionData(functionFragment: "unpauser", values?: undefined): string;

  decodeFunctionResult(functionFragment: "pauser", data: BytesLike): Result;
  decodeFunctionResult(functionFragment: "unpauser", data: BytesLike): Result;

  events: {};
}

export interface IPauserRegistry extends BaseContract {
  connect(signerOrProvider: Signer | Provider | string): this;
  attach(addressOrName: string): this;
  deployed(): Promise<this>;

  interface: IPauserRegistryInterface;

  queryFilter<TEvent extends TypedEvent>(
    event: TypedEventFilter<TEvent>,
    fromBlockOrBlockhash?: string | number | undefined,
    toBlock?: string | number | undefined
  ): Promise<Array<TEvent>>;

  listeners<TEvent extends TypedEvent>(
    eventFilter?: TypedEventFilter<TEvent>
  ): Array<TypedListener<TEvent>>;
  listeners(eventName?: string): Array<Listener>;
  removeAllListeners<TEvent extends TypedEvent>(
    eventFilter: TypedEventFilter<TEvent>
  ): this;
  removeAllListeners(eventName?: string): this;
  off: OnEvent<this>;
  on: OnEvent<this>;
  once: OnEvent<this>;
  removeListener: OnEvent<this>;

  functions: {
    pauser(overrides?: CallOverrides): Promise<[string]>;

    unpauser(overrides?: CallOverrides): Promise<[string]>;
  };

  pauser(overrides?: CallOverrides): Promise<string>;

  unpauser(overrides?: CallOverrides): Promise<string>;

  callStatic: {
    pauser(overrides?: CallOverrides): Promise<string>;

    unpauser(overrides?: CallOverrides): Promise<string>;
  };

  filters: {};

  estimateGas: {
    pauser(overrides?: CallOverrides): Promise<BigNumber>;

    unpauser(overrides?: CallOverrides): Promise<BigNumber>;
  };

  populateTransaction: {
    pauser(overrides?: CallOverrides): Promise<PopulatedTransaction>;

    unpauser(overrides?: CallOverrides): Promise<PopulatedTransaction>;
  };
}
