// @mantleio/contracts/dist/types/contracts/libraries/eigenda/interfaces/IQuorumRegistry.ts
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
import type {
  BaseContract,
  BigNumber,
  BigNumberish,
  BytesLike,
  CallOverrides,
  PopulatedTransaction,
  Signer,
  utils,
} from "ethers";
import type { FunctionFragment, Result } from "@ethersproject/abi";
import type { Listener, Provider } from "@ethersproject/providers";
import type {
  TypedEventFilter,
  TypedEvent,
  TypedListener,
  OnEvent,
  PromiseOrValue,
} from "../../../../common";

export declare namespace IQuorumRegistry {
  export type OperatorStakeStruct = {
    updateBlockNumber: PromiseOrValue<BigNumberish>;
    nextUpdateBlockNumber: PromiseOrValue<BigNumberish>;
    firstQuorumStake: PromiseOrValue<BigNumberish>;
    secondQuorumStake: PromiseOrValue<BigNumberish>;
  };

  export type OperatorStakeStructOutput = [
    number,
    number,
    BigNumber,
    BigNumber
  ] & {
    updateBlockNumber: number;
    nextUpdateBlockNumber: number;
    firstQuorumStake: BigNumber;
    secondQuorumStake: BigNumber;
  };
}

export interface IQuorumRegistryInterface extends utils.Interface {
  functions: {
    "checkOperatorActiveAtBlockNumber(address,uint256,uint256)": FunctionFragment;
    "checkOperatorInactiveAtBlockNumber(address,uint256,uint256)": FunctionFragment;
    "getFromTaskNumberForOperator(address)": FunctionFragment;
    "getLengthOfTotalStakeHistory()": FunctionFragment;
    "getOperatorIndex(address,uint32,uint32)": FunctionFragment;
    "getOperatorPubkeyHash(address)": FunctionFragment;
    "getStakeFromPubkeyHashAndIndex(bytes32,uint256)": FunctionFragment;
    "getTotalOperators(uint32,uint32)": FunctionFragment;
    "getTotalStakeFromIndex(uint256)": FunctionFragment;
    "isActiveOperator(address)": FunctionFragment;
    "numOperators()": FunctionFragment;
    "operatorStakes(address)": FunctionFragment;
    "totalStake()": FunctionFragment;
  };

  getFunction(
    nameOrSignatureOrTopic:
      | "checkOperatorActiveAtBlockNumber"
      | "checkOperatorInactiveAtBlockNumber"
      | "getFromTaskNumberForOperator"
      | "getLengthOfTotalStakeHistory"
      | "getOperatorIndex"
      | "getOperatorPubkeyHash"
      | "getStakeFromPubkeyHashAndIndex"
      | "getTotalOperators"
      | "getTotalStakeFromIndex"
      | "isActiveOperator"
      | "numOperators"
      | "operatorStakes"
      | "totalStake"
  ): FunctionFragment;

  encodeFunctionData(
    functionFragment: "checkOperatorActiveAtBlockNumber",
    values: [
      PromiseOrValue<string>,
      PromiseOrValue<BigNumberish>,
      PromiseOrValue<BigNumberish>
    ]
  ): string;
  encodeFunctionData(
    functionFragment: "checkOperatorInactiveAtBlockNumber",
    values: [
      PromiseOrValue<string>,
      PromiseOrValue<BigNumberish>,
      PromiseOrValue<BigNumberish>
    ]
  ): string;
  encodeFunctionData(
    functionFragment: "getFromTaskNumberForOperator",
    values: [PromiseOrValue<string>]
  ): string;
  encodeFunctionData(
    functionFragment: "getLengthOfTotalStakeHistory",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "getOperatorIndex",
    values: [
      PromiseOrValue<string>,
      PromiseOrValue<BigNumberish>,
      PromiseOrValue<BigNumberish>
    ]
  ): string;
  encodeFunctionData(
    functionFragment: "getOperatorPubkeyHash",
    values: [PromiseOrValue<string>]
  ): string;
  encodeFunctionData(
    functionFragment: "getStakeFromPubkeyHashAndIndex",
    values: [PromiseOrValue<BytesLike>, PromiseOrValue<BigNumberish>]
  ): string;
  encodeFunctionData(
    functionFragment: "getTotalOperators",
    values: [PromiseOrValue<BigNumberish>, PromiseOrValue<BigNumberish>]
  ): string;
  encodeFunctionData(
    functionFragment: "getTotalStakeFromIndex",
    values: [PromiseOrValue<BigNumberish>]
  ): string;
  encodeFunctionData(
    functionFragment: "isActiveOperator",
    values: [PromiseOrValue<string>]
  ): string;
  encodeFunctionData(
    functionFragment: "numOperators",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "operatorStakes",
    values: [PromiseOrValue<string>]
  ): string;
  encodeFunctionData(
    functionFragment: "totalStake",
    values?: undefined
  ): string;

  decodeFunctionResult(
    functionFragment: "checkOperatorActiveAtBlockNumber",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "checkOperatorInactiveAtBlockNumber",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "getFromTaskNumberForOperator",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "getLengthOfTotalStakeHistory",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "getOperatorIndex",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "getOperatorPubkeyHash",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "getStakeFromPubkeyHashAndIndex",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "getTotalOperators",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "getTotalStakeFromIndex",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "isActiveOperator",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "numOperators",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "operatorStakes",
    data: BytesLike
  ): Result;
  decodeFunctionResult(functionFragment: "totalStake", data: BytesLike): Result;

  events: {};
}

export interface IQuorumRegistry extends BaseContract {
  connect(signerOrProvider: Signer | Provider | string): this;
  attach(addressOrName: string): this;
  deployed(): Promise<this>;

  interface: IQuorumRegistryInterface;

  queryFilter<TEvent extends TypedEvent>(
    event: TypedEventFilter<TEvent>,
    fromBlockOrBlockhash?: string | number | undefined,
    toBlock?: string | number | undefined
  ): Promise<Array<TEvent>>;

  listeners<TEvent extends TypedEvent>(
    eventFilter?: TypedEventFilter<TEvent>
  ): Array<TypedListener<TEvent>>;
  listeners(eventName?: string): Array<Listener>;
  removeAllListeners<TEvent extends TypedEvent>(
    eventFilter: TypedEventFilter<TEvent>
  ): this;
  removeAllListeners(eventName?: string): this;
  off: OnEvent<this>;
  on: OnEvent<this>;
  once: OnEvent<this>;
  removeListener: OnEvent<this>;

  functions: {
    checkOperatorActiveAtBlockNumber(
      operator: PromiseOrValue<string>,
      blockNumber: PromiseOrValue<BigNumberish>,
      stakeHistoryIndex: PromiseOrValue<BigNumberish>,
      overrides?: CallOverrides
    ): Promise<[boolean]>;

    checkOperatorInactiveAtBlockNumber(
      operator: PromiseOrValue<string>,
      blockNumber: PromiseOrValue<BigNumberish>,
      stakeHistoryIndex: PromiseOrValue<BigNumberish>,
      overrides?: CallOverrides
    ): Promise<[boolean]>;

    getFromTaskNumberForOperator(
      operator: PromiseOrValue<string>,
      overrides?: CallOverrides
    ): Promise<[number]>;

    getLengthOfTotalStakeHistory(
      overrides?: CallOverrides
    ): Promise<[BigNumber]>;

    getOperatorIndex(
      operator: PromiseOrValue<string>,
      blockNumber: PromiseOrValue<BigNumberish>,
      index: PromiseOrValue<BigNumberish>,
      overrides?: CallOverrides
    ): Promise<[number]>;

    getOperatorPubkeyHash(
      operator: PromiseOrValue<string>,
      overrides?: CallOverrides
    ): Promise<[string]>;

    getStakeFromPubkeyHashAndIndex(
      pubkeyHash: PromiseOrValue<BytesLike>,
      index: PromiseOrValue<BigNumberish>,
      overrides?: CallOverrides
    ): Promise<[IQuorumRegistry.OperatorStakeStructOutput]>;

    getTotalOperators(
      blockNumber: PromiseOrValue<BigNumberish>,
      index: PromiseOrValue<BigNumberish>,
      overrides?: CallOverrides
    ): Promise<[number]>;

    getTotalStakeFromIndex(
      index: PromiseOrValue<BigNumberish>,
      overrides?: CallOverrides
    ): Promise<[IQuorumRegistry.OperatorStakeStructOutput]>;

    isActiveOperator(
      operator: PromiseOrValue<string>,
      overrides?: CallOverrides
    ): Promise<[boolean]>;

    numOperators(overrides?: CallOverrides): Promise<[number]>;

    operatorStakes(
      operator: PromiseOrValue<string>,
      overrides?: CallOverrides
    ): Promise<[BigNumber, BigNumber]>;

    totalStake(overrides?: CallOverrides): Promise<[BigNumber, BigNumber]>;
  };

  checkOperatorActiveAtBlockNumber(
    operator: PromiseOrValue<string>,
    blockNumber: PromiseOrValue<BigNumberish>,
    stakeHistoryIndex: PromiseOrValue<BigNumberish>,
    overrides?: CallOverrides
  ): Promise<boolean>;

  checkOperatorInactiveAtBlockNumber(
    operator: PromiseOrValue<string>,
    blockNumber: PromiseOrValue<BigNumberish>,
    stakeHistoryIndex: PromiseOrValue<BigNumberish>,
    overrides?: CallOverrides
  ): Promise<boolean>;

  getFromTaskNumberForOperator(
    operator: PromiseOrValue<string>,
    overrides?: CallOverrides
  ): Promise<number>;

  getLengthOfTotalStakeHistory(overrides?: CallOverrides): Promise<BigNumber>;

  getOperatorIndex(
    operator: PromiseOrValue<string>,
    blockNumber: PromiseOrValue<BigNumberish>,
    index: PromiseOrValue<BigNumberish>,
    overrides?: CallOverrides
  ): Promise<number>;

  getOperatorPubkeyHash(
    operator: PromiseOrValue<string>,
    overrides?: CallOverrides
  ): Promise<string>;

  getStakeFromPubkeyHashAndIndex(
    pubkeyHash: PromiseOrValue<BytesLike>,
    index: PromiseOrValue<BigNumberish>,
    overrides?: CallOverrides
  ): Promise<IQuorumRegistry.OperatorStakeStructOutput>;

  getTotalOperators(
    blockNumber: PromiseOrValue<BigNumberish>,
    index: PromiseOrValue<BigNumberish>,
    overrides?: CallOverrides
  ): Promise<number>;

  getTotalStakeFromIndex(
    index: PromiseOrValue<BigNumberish>,
    overrides?: CallOverrides
  ): Promise<IQuorumRegistry.OperatorStakeStructOutput>;

  isActiveOperator(
    operator: PromiseOrValue<string>,
    overrides?: CallOverrides
  ): Promise<boolean>;

  numOperators(overrides?: CallOverrides): Promise<number>;

  operatorStakes(
    operator: PromiseOrValue<string>,
    overrides?: CallOverrides
  ): Promise<[BigNumber, BigNumber]>;

  totalStake(overrides?: CallOverrides): Promise<[BigNumber, BigNumber]>;

  callStatic: {
    checkOperatorActiveAtBlockNumber(
      operator: PromiseOrValue<string>,
      blockNumber: PromiseOrValue<BigNumberish>,
      stakeHistoryIndex: PromiseOrValue<BigNumberish>,
      overrides?: CallOverrides
    ): Promise<boolean>;

    checkOperatorInactiveAtBlockNumber(
      operator: PromiseOrValue<string>,
      blockNumber: PromiseOrValue<BigNumberish>,
      stakeHistoryIndex: PromiseOrValue<BigNumberish>,
      overrides?: CallOverrides
    ): Promise<boolean>;

    getFromTaskNumberForOperator(
      operator: PromiseOrValue<string>,
      overrides?: CallOverrides
    ): Promise<number>;

    getLengthOfTotalStakeHistory(overrides?: CallOverrides): Promise<BigNumber>;

    getOperatorIndex(
      operator: PromiseOrValue<string>,
      blockNumber: PromiseOrValue<BigNumberish>,
      index: PromiseOrValue<BigNumberish>,
      overrides?: CallOverrides
    ): Promise<number>;

    getOperatorPubkeyHash(
      operator: PromiseOrValue<string>,
      overrides?: CallOverrides
    ): Promise<string>;

    getStakeFromPubkeyHashAndIndex(
      pubkeyHash: PromiseOrValue<BytesLike>,
      index: PromiseOrValue<BigNumberish>,
      overrides?: CallOverrides
    ): Promise<IQuorumRegistry.OperatorStakeStructOutput>;

    getTotalOperators(
      blockNumber: PromiseOrValue<BigNumberish>,
      index: PromiseOrValue<BigNumberish>,
      overrides?: CallOverrides
    ): Promise<number>;

    getTotalStakeFromIndex(
      index: PromiseOrValue<BigNumberish>,
      overrides?: CallOverrides
    ): Promise<IQuorumRegistry.OperatorStakeStructOutput>;

    isActiveOperator(
      operator: PromiseOrValue<string>,
      overrides?: CallOverrides
    ): Promise<boolean>;

    numOperators(overrides?: CallOverrides): Promise<number>;

    operatorStakes(
      operator: PromiseOrValue<string>,
      overrides?: CallOverrides
    ): Promise<[BigNumber, BigNumber]>;

    totalStake(overrides?: CallOverrides): Promise<[BigNumber, BigNumber]>;
  };

  filters: {};

  estimateGas: {
    checkOperatorActiveAtBlockNumber(
      operator: PromiseOrValue<string>,
      blockNumber: PromiseOrValue<BigNumberish>,
      stakeHistoryIndex: PromiseOrValue<BigNumberish>,
      overrides?: CallOverrides
    ): Promise<BigNumber>;

    checkOperatorInactiveAtBlockNumber(
      operator: PromiseOrValue<string>,
      blockNumber: PromiseOrValue<BigNumberish>,
      stakeHistoryIndex: PromiseOrValue<BigNumberish>,
      overrides?: CallOverrides
    ): Promise<BigNumber>;

    getFromTaskNumberForOperator(
      operator: PromiseOrValue<string>,
      overrides?: CallOverrides
    ): Promise<BigNumber>;

    getLengthOfTotalStakeHistory(overrides?: CallOverrides): Promise<BigNumber>;

    getOperatorIndex(
      operator: PromiseOrValue<string>,
      blockNumber: PromiseOrValue<BigNumberish>,
      index: PromiseOrValue<BigNumberish>,
      overrides?: CallOverrides
    ): Promise<BigNumber>;

    getOperatorPubkeyHash(
      operator: PromiseOrValue<string>,
      overrides?: CallOverrides
    ): Promise<BigNumber>;

    getStakeFromPubkeyHashAndIndex(
      pubkeyHash: PromiseOrValue<BytesLike>,
      index: PromiseOrValue<BigNumberish>,
      overrides?: CallOverrides
    ): Promise<BigNumber>;

    getTotalOperators(
      blockNumber: PromiseOrValue<BigNumberish>,
      index: PromiseOrValue<BigNumberish>,
      overrides?: CallOverrides
    ): Promise<BigNumber>;

    getTotalStakeFromIndex(
      index: PromiseOrValue<BigNumberish>,
      overrides?: CallOverrides
    ): Promise<BigNumber>;

    isActiveOperator(
      operator: PromiseOrValue<string>,
      overrides?: CallOverrides
    ): Promise<BigNumber>;

    numOperators(overrides?: CallOverrides): Promise<BigNumber>;

    operatorStakes(
      operator: PromiseOrValue<string>,
      overrides?: CallOverrides
    ): Promise<BigNumber>;

    totalStake(overrides?: CallOverrides): Promise<BigNumber>;
  };

  populateTransaction: {
    checkOperatorActiveAtBlockNumber(
      operator: PromiseOrValue<string>,
      blockNumber: PromiseOrValue<BigNumberish>,
      stakeHistoryIndex: PromiseOrValue<BigNumberish>,
      overrides?: CallOverrides
    ): Promise<PopulatedTransaction>;

    checkOperatorInactiveAtBlockNumber(
      operator: PromiseOrValue<string>,
      blockNumber: PromiseOrValue<BigNumberish>,
      stakeHistoryIndex: PromiseOrValue<BigNumberish>,
      overrides?: CallOverrides
    ): Promise<PopulatedTransaction>;

    getFromTaskNumberForOperator(
      operator: PromiseOrValue<string>,
      overrides?: CallOverrides
    ): Promise<PopulatedTransaction>;

    getLengthOfTotalStakeHistory(
      overrides?: CallOverrides
    ): Promise<PopulatedTransaction>;

    getOperatorIndex(
      operator: PromiseOrValue<string>,
      blockNumber: PromiseOrValue<BigNumberish>,
      index: PromiseOrValue<BigNumberish>,
      overrides?: CallOverrides
    ): Promise<PopulatedTransaction>;

    getOperatorPubkeyHash(
      operator: PromiseOrValue<string>,
      overrides?: CallOverrides
    ): Promise<PopulatedTransaction>;

    getStakeFromPubkeyHashAndIndex(
      pubkeyHash: PromiseOrValue<BytesLike>,
      index: PromiseOrValue<BigNumberish>,
      overrides?: CallOverrides
    ): Promise<PopulatedTransaction>;

    getTotalOperators(
      blockNumber: PromiseOrValue<BigNumberish>,
      index: PromiseOrValue<BigNumberish>,
      overrides?: CallOverrides
    ): Promise<PopulatedTransaction>;

    getTotalStakeFromIndex(
      index: PromiseOrValue<BigNumberish>,
      overrides?: CallOverrides
    ): Promise<PopulatedTransaction>;

    isActiveOperator(
      operator: PromiseOrValue<string>,
      overrides?: CallOverrides
    ): Promise<PopulatedTransaction>;

    numOperators(overrides?: CallOverrides): Promise<PopulatedTransaction>;

    operatorStakes(
      operator: PromiseOrValue<string>,
      overrides?: CallOverrides
    ): Promise<PopulatedTransaction>;

    totalStake(overrides?: CallOverrides): Promise<PopulatedTransaction>;
  };
}
