// @mantleio/contracts/dist/types/contracts/libraries/eigenda/interfaces/ISlasher.ts
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
import type {
  BaseContract,
  BigNumber,
  BigNumberish,
  BytesLike,
  CallOverrides,
  ContractTransaction,
  Overrides,
  PopulatedTransaction,
  Signer,
  utils,
} from "ethers";
import type { FunctionFragment, Result } from "@ethersproject/abi";
import type { Listener, Provider } from "@ethersproject/providers";
import type {
  TypedEventFilter,
  TypedEvent,
  TypedListener,
  OnEvent,
  PromiseOrValue,
} from "../../../../common";

export interface ISlasherInterface extends utils.Interface {
  functions: {
    "bondedUntil(address,address)": FunctionFragment;
    "canSlash(address,address)": FunctionFragment;
    "canWithdraw(address,uint32,uint256)": FunctionFragment;
    "freezeOperator(address)": FunctionFragment;
    "getCorrectValueForInsertAfter(address,uint32)": FunctionFragment;
    "getMiddlewareTimesIndexBlock(address,uint32)": FunctionFragment;
    "getMiddlewareTimesIndexServeUntil(address,uint32)": FunctionFragment;
    "isFrozen(address)": FunctionFragment;
    "latestUpdateBlock(address,address)": FunctionFragment;
    "optIntoSlashing(address)": FunctionFragment;
    "recordFirstStakeUpdate(address,uint32)": FunctionFragment;
    "recordLastStakeUpdateAndRevokeSlashingAbility(address,uint32)": FunctionFragment;
    "recordStakeUpdate(address,uint32,uint32,uint256)": FunctionFragment;
    "resetFrozenStatus(address[])": FunctionFragment;
  };

  getFunction(
    nameOrSignatureOrTopic:
      | "bondedUntil"
      | "canSlash"
      | "canWithdraw"
      | "freezeOperator"
      | "getCorrectValueForInsertAfter"
      | "getMiddlewareTimesIndexBlock"
      | "getMiddlewareTimesIndexServeUntil"
      | "isFrozen"
      | "latestUpdateBlock"
      | "optIntoSlashing"
      | "recordFirstStakeUpdate"
      | "recordLastStakeUpdateAndRevokeSlashingAbility"
      | "recordStakeUpdate"
      | "resetFrozenStatus"
  ): FunctionFragment;

  encodeFunctionData(
    functionFragment: "bondedUntil",
    values: [PromiseOrValue<string>, PromiseOrValue<string>]
  ): string;
  encodeFunctionData(
    functionFragment: "canSlash",
    values: [PromiseOrValue<string>, PromiseOrValue<string>]
  ): string;
  encodeFunctionData(
    functionFragment: "canWithdraw",
    values: [
      PromiseOrValue<string>,
      PromiseOrValue<BigNumberish>,
      PromiseOrValue<BigNumberish>
    ]
  ): string;
  encodeFunctionData(
    functionFragment: "freezeOperator",
    values: [PromiseOrValue<string>]
  ): string;
  encodeFunctionData(
    functionFragment: "getCorrectValueForInsertAfter",
    values: [PromiseOrValue<string>, PromiseOrValue<BigNumberish>]
  ): string;
  encodeFunctionData(
    functionFragment: "getMiddlewareTimesIndexBlock",
    values: [PromiseOrValue<string>, PromiseOrValue<BigNumberish>]
  ): string;
  encodeFunctionData(
    functionFragment: "getMiddlewareTimesIndexServeUntil",
    values: [PromiseOrValue<string>, PromiseOrValue<BigNumberish>]
  ): string;
  encodeFunctionData(
    functionFragment: "isFrozen",
    values: [PromiseOrValue<string>]
  ): string;
  encodeFunctionData(
    functionFragment: "latestUpdateBlock",
    values: [PromiseOrValue<string>, PromiseOrValue<string>]
  ): string;
  encodeFunctionData(
    functionFragment: "optIntoSlashing",
    values: [PromiseOrValue<string>]
  ): string;
  encodeFunctionData(
    functionFragment: "recordFirstStakeUpdate",
    values: [PromiseOrValue<string>, PromiseOrValue<BigNumberish>]
  ): string;
  encodeFunctionData(
    functionFragment: "recordLastStakeUpdateAndRevokeSlashingAbility",
    values: [PromiseOrValue<string>, PromiseOrValue<BigNumberish>]
  ): string;
  encodeFunctionData(
    functionFragment: "recordStakeUpdate",
    values: [
      PromiseOrValue<string>,
      PromiseOrValue<BigNumberish>,
      PromiseOrValue<BigNumberish>,
      PromiseOrValue<BigNumberish>
    ]
  ): string;
  encodeFunctionData(
    functionFragment: "resetFrozenStatus",
    values: [PromiseOrValue<string>[]]
  ): string;

  decodeFunctionResult(
    functionFragment: "bondedUntil",
    data: BytesLike
  ): Result;
  decodeFunctionResult(functionFragment: "canSlash", data: BytesLike): Result;
  decodeFunctionResult(
    functionFragment: "canWithdraw",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "freezeOperator",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "getCorrectValueForInsertAfter",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "getMiddlewareTimesIndexBlock",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "getMiddlewareTimesIndexServeUntil",
    data: BytesLike
  ): Result;
  decodeFunctionResult(functionFragment: "isFrozen", data: BytesLike): Result;
  decodeFunctionResult(
    functionFragment: "latestUpdateBlock",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "optIntoSlashing",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "recordFirstStakeUpdate",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "recordLastStakeUpdateAndRevokeSlashingAbility",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "recordStakeUpdate",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "resetFrozenStatus",
    data: BytesLike
  ): Result;

  events: {};
}

export interface ISlasher extends BaseContract {
  connect(signerOrProvider: Signer | Provider | string): this;
  attach(addressOrName: string): this;
  deployed(): Promise<this>;

  interface: ISlasherInterface;

  queryFilter<TEvent extends TypedEvent>(
    event: TypedEventFilter<TEvent>,
    fromBlockOrBlockhash?: string | number | undefined,
    toBlock?: string | number | undefined
  ): Promise<Array<TEvent>>;

  listeners<TEvent extends TypedEvent>(
    eventFilter?: TypedEventFilter<TEvent>
  ): Array<TypedListener<TEvent>>;
  listeners(eventName?: string): Array<Listener>;
  removeAllListeners<TEvent extends TypedEvent>(
    eventFilter: TypedEventFilter<TEvent>
  ): this;
  removeAllListeners(eventName?: string): this;
  off: OnEvent<this>;
  on: OnEvent<this>;
  once: OnEvent<this>;
  removeListener: OnEvent<this>;

  functions: {
    bondedUntil(
      operator: PromiseOrValue<string>,
      serviceContract: PromiseOrValue<string>,
      overrides?: CallOverrides
    ): Promise<[number]>;

    canSlash(
      toBeSlashed: PromiseOrValue<string>,
      slashingContract: PromiseOrValue<string>,
      overrides?: CallOverrides
    ): Promise<[boolean]>;

    canWithdraw(
      operator: PromiseOrValue<string>,
      withdrawalStartBlock: PromiseOrValue<BigNumberish>,
      middlewareTimesIndex: PromiseOrValue<BigNumberish>,
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<ContractTransaction>;

    freezeOperator(
      toBeFrozen: PromiseOrValue<string>,
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<ContractTransaction>;

    getCorrectValueForInsertAfter(
      operator: PromiseOrValue<string>,
      updateBlock: PromiseOrValue<BigNumberish>,
      overrides?: CallOverrides
    ): Promise<[BigNumber]>;

    getMiddlewareTimesIndexBlock(
      operator: PromiseOrValue<string>,
      index: PromiseOrValue<BigNumberish>,
      overrides?: CallOverrides
    ): Promise<[number]>;

    getMiddlewareTimesIndexServeUntil(
      operator: PromiseOrValue<string>,
      index: PromiseOrValue<BigNumberish>,
      overrides?: CallOverrides
    ): Promise<[number]>;

    isFrozen(
      staker: PromiseOrValue<string>,
      overrides?: CallOverrides
    ): Promise<[boolean]>;

    latestUpdateBlock(
      operator: PromiseOrValue<string>,
      serviceContract: PromiseOrValue<string>,
      overrides?: CallOverrides
    ): Promise<[number]>;

    optIntoSlashing(
      contractAddress: PromiseOrValue<string>,
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<ContractTransaction>;

    recordFirstStakeUpdate(
      operator: PromiseOrValue<string>,
      serveUntil: PromiseOrValue<BigNumberish>,
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<ContractTransaction>;

    recordLastStakeUpdateAndRevokeSlashingAbility(
      operator: PromiseOrValue<string>,
      serveUntil: PromiseOrValue<BigNumberish>,
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<ContractTransaction>;

    recordStakeUpdate(
      operator: PromiseOrValue<string>,
      updateBlock: PromiseOrValue<BigNumberish>,
      serveUntil: PromiseOrValue<BigNumberish>,
      insertAfter: PromiseOrValue<BigNumberish>,
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<ContractTransaction>;

    resetFrozenStatus(
      frozenAddresses: PromiseOrValue<string>[],
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<ContractTransaction>;
  };

  bondedUntil(
    operator: PromiseOrValue<string>,
    serviceContract: PromiseOrValue<string>,
    overrides?: CallOverrides
  ): Promise<number>;

  canSlash(
    toBeSlashed: PromiseOrValue<string>,
    slashingContract: PromiseOrValue<string>,
    overrides?: CallOverrides
  ): Promise<boolean>;

  canWithdraw(
    operator: PromiseOrValue<string>,
    withdrawalStartBlock: PromiseOrValue<BigNumberish>,
    middlewareTimesIndex: PromiseOrValue<BigNumberish>,
    overrides?: Overrides & { from?: PromiseOrValue<string> }
  ): Promise<ContractTransaction>;

  freezeOperator(
    toBeFrozen: PromiseOrValue<string>,
    overrides?: Overrides & { from?: PromiseOrValue<string> }
  ): Promise<ContractTransaction>;

  getCorrectValueForInsertAfter(
    operator: PromiseOrValue<string>,
    updateBlock: PromiseOrValue<BigNumberish>,
    overrides?: CallOverrides
  ): Promise<BigNumber>;

  getMiddlewareTimesIndexBlock(
    operator: PromiseOrValue<string>,
    index: PromiseOrValue<BigNumberish>,
    overrides?: CallOverrides
  ): Promise<number>;

  getMiddlewareTimesIndexServeUntil(
    operator: PromiseOrValue<string>,
    index: PromiseOrValue<BigNumberish>,
    overrides?: CallOverrides
  ): Promise<number>;

  isFrozen(
    staker: PromiseOrValue<string>,
    overrides?: CallOverrides
  ): Promise<boolean>;

  latestUpdateBlock(
    operator: PromiseOrValue<string>,
    serviceContract: PromiseOrValue<string>,
    overrides?: CallOverrides
  ): Promise<number>;

  optIntoSlashing(
    contractAddress: PromiseOrValue<string>,
    overrides?: Overrides & { from?: PromiseOrValue<string> }
  ): Promise<ContractTransaction>;

  recordFirstStakeUpdate(
    operator: PromiseOrValue<string>,
    serveUntil: PromiseOrValue<BigNumberish>,
    overrides?: Overrides & { from?: PromiseOrValue<string> }
  ): Promise<ContractTransaction>;

  recordLastStakeUpdateAndRevokeSlashingAbility(
    operator: PromiseOrValue<string>,
    serveUntil: PromiseOrValue<BigNumberish>,
    overrides?: Overrides & { from?: PromiseOrValue<string> }
  ): Promise<ContractTransaction>;

  recordStakeUpdate(
    operator: PromiseOrValue<string>,
    updateBlock: PromiseOrValue<BigNumberish>,
    serveUntil: PromiseOrValue<BigNumberish>,
    insertAfter: PromiseOrValue<BigNumberish>,
    overrides?: Overrides & { from?: PromiseOrValue<string> }
  ): Promise<ContractTransaction>;

  resetFrozenStatus(
    frozenAddresses: PromiseOrValue<string>[],
    overrides?: Overrides & { from?: PromiseOrValue<string> }
  ): Promise<ContractTransaction>;

  callStatic: {
    bondedUntil(
      operator: PromiseOrValue<string>,
      serviceContract: PromiseOrValue<string>,
      overrides?: CallOverrides
    ): Promise<number>;

    canSlash(
      toBeSlashed: PromiseOrValue<string>,
      slashingContract: PromiseOrValue<string>,
      overrides?: CallOverrides
    ): Promise<boolean>;

    canWithdraw(
      operator: PromiseOrValue<string>,
      withdrawalStartBlock: PromiseOrValue<BigNumberish>,
      middlewareTimesIndex: PromiseOrValue<BigNumberish>,
      overrides?: CallOverrides
    ): Promise<boolean>;

    freezeOperator(
      toBeFrozen: PromiseOrValue<string>,
      overrides?: CallOverrides
    ): Promise<void>;

    getCorrectValueForInsertAfter(
      operator: PromiseOrValue<string>,
      updateBlock: PromiseOrValue<BigNumberish>,
      overrides?: CallOverrides
    ): Promise<BigNumber>;

    getMiddlewareTimesIndexBlock(
      operator: PromiseOrValue<string>,
      index: PromiseOrValue<BigNumberish>,
      overrides?: CallOverrides
    ): Promise<number>;

    getMiddlewareTimesIndexServeUntil(
      operator: PromiseOrValue<string>,
      index: PromiseOrValue<BigNumberish>,
      overrides?: CallOverrides
    ): Promise<number>;

    isFrozen(
      staker: PromiseOrValue<string>,
      overrides?: CallOverrides
    ): Promise<boolean>;

    latestUpdateBlock(
      operator: PromiseOrValue<string>,
      serviceContract: PromiseOrValue<string>,
      overrides?: CallOverrides
    ): Promise<number>;

    optIntoSlashing(
      contractAddress: PromiseOrValue<string>,
      overrides?: CallOverrides
    ): Promise<void>;

    recordFirstStakeUpdate(
      operator: PromiseOrValue<string>,
      serveUntil: PromiseOrValue<BigNumberish>,
      overrides?: CallOverrides
    ): Promise<void>;

    recordLastStakeUpdateAndRevokeSlashingAbility(
      operator: PromiseOrValue<string>,
      serveUntil: PromiseOrValue<BigNumberish>,
      overrides?: CallOverrides
    ): Promise<void>;

    recordStakeUpdate(
      operator: PromiseOrValue<string>,
      updateBlock: PromiseOrValue<BigNumberish>,
      serveUntil: PromiseOrValue<BigNumberish>,
      insertAfter: PromiseOrValue<BigNumberish>,
      overrides?: CallOverrides
    ): Promise<void>;

    resetFrozenStatus(
      frozenAddresses: PromiseOrValue<string>[],
      overrides?: CallOverrides
    ): Promise<void>;
  };

  filters: {};

  estimateGas: {
    bondedUntil(
      operator: PromiseOrValue<string>,
      serviceContract: PromiseOrValue<string>,
      overrides?: CallOverrides
    ): Promise<BigNumber>;

    canSlash(
      toBeSlashed: PromiseOrValue<string>,
      slashingContract: PromiseOrValue<string>,
      overrides?: CallOverrides
    ): Promise<BigNumber>;

    canWithdraw(
      operator: PromiseOrValue<string>,
      withdrawalStartBlock: PromiseOrValue<BigNumberish>,
      middlewareTimesIndex: PromiseOrValue<BigNumberish>,
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<BigNumber>;

    freezeOperator(
      toBeFrozen: PromiseOrValue<string>,
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<BigNumber>;

    getCorrectValueForInsertAfter(
      operator: PromiseOrValue<string>,
      updateBlock: PromiseOrValue<BigNumberish>,
      overrides?: CallOverrides
    ): Promise<BigNumber>;

    getMiddlewareTimesIndexBlock(
      operator: PromiseOrValue<string>,
      index: PromiseOrValue<BigNumberish>,
      overrides?: CallOverrides
    ): Promise<BigNumber>;

    getMiddlewareTimesIndexServeUntil(
      operator: PromiseOrValue<string>,
      index: PromiseOrValue<BigNumberish>,
      overrides?: CallOverrides
    ): Promise<BigNumber>;

    isFrozen(
      staker: PromiseOrValue<string>,
      overrides?: CallOverrides
    ): Promise<BigNumber>;

    latestUpdateBlock(
      operator: PromiseOrValue<string>,
      serviceContract: PromiseOrValue<string>,
      overrides?: CallOverrides
    ): Promise<BigNumber>;

    optIntoSlashing(
      contractAddress: PromiseOrValue<string>,
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<BigNumber>;

    recordFirstStakeUpdate(
      operator: PromiseOrValue<string>,
      serveUntil: PromiseOrValue<BigNumberish>,
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<BigNumber>;

    recordLastStakeUpdateAndRevokeSlashingAbility(
      operator: PromiseOrValue<string>,
      serveUntil: PromiseOrValue<BigNumberish>,
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<BigNumber>;

    recordStakeUpdate(
      operator: PromiseOrValue<string>,
      updateBlock: PromiseOrValue<BigNumberish>,
      serveUntil: PromiseOrValue<BigNumberish>,
      insertAfter: PromiseOrValue<BigNumberish>,
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<BigNumber>;

    resetFrozenStatus(
      frozenAddresses: PromiseOrValue<string>[],
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<BigNumber>;
  };

  populateTransaction: {
    bondedUntil(
      operator: PromiseOrValue<string>,
      serviceContract: PromiseOrValue<string>,
      overrides?: CallOverrides
    ): Promise<PopulatedTransaction>;

    canSlash(
      toBeSlashed: PromiseOrValue<string>,
      slashingContract: PromiseOrValue<string>,
      overrides?: CallOverrides
    ): Promise<PopulatedTransaction>;

    canWithdraw(
      operator: PromiseOrValue<string>,
      withdrawalStartBlock: PromiseOrValue<BigNumberish>,
      middlewareTimesIndex: PromiseOrValue<BigNumberish>,
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<PopulatedTransaction>;

    freezeOperator(
      toBeFrozen: PromiseOrValue<string>,
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<PopulatedTransaction>;

    getCorrectValueForInsertAfter(
      operator: PromiseOrValue<string>,
      updateBlock: PromiseOrValue<BigNumberish>,
      overrides?: CallOverrides
    ): Promise<PopulatedTransaction>;

    getMiddlewareTimesIndexBlock(
      operator: PromiseOrValue<string>,
      index: PromiseOrValue<BigNumberish>,
      overrides?: CallOverrides
    ): Promise<PopulatedTransaction>;

    getMiddlewareTimesIndexServeUntil(
      operator: PromiseOrValue<string>,
      index: PromiseOrValue<BigNumberish>,
      overrides?: CallOverrides
    ): Promise<PopulatedTransaction>;

    isFrozen(
      staker: PromiseOrValue<string>,
      overrides?: CallOverrides
    ): Promise<PopulatedTransaction>;

    latestUpdateBlock(
      operator: PromiseOrValue<string>,
      serviceContract: PromiseOrValue<string>,
      overrides?: CallOverrides
    ): Promise<PopulatedTransaction>;

    optIntoSlashing(
      contractAddress: PromiseOrValue<string>,
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<PopulatedTransaction>;

    recordFirstStakeUpdate(
      operator: PromiseOrValue<string>,
      serveUntil: PromiseOrValue<BigNumberish>,
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<PopulatedTransaction>;

    recordLastStakeUpdateAndRevokeSlashingAbility(
      operator: PromiseOrValue<string>,
      serveUntil: PromiseOrValue<BigNumberish>,
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<PopulatedTransaction>;

    recordStakeUpdate(
      operator: PromiseOrValue<string>,
      updateBlock: PromiseOrValue<BigNumberish>,
      serveUntil: PromiseOrValue<BigNumberish>,
      insertAfter: PromiseOrValue<BigNumberish>,
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<PopulatedTransaction>;

    resetFrozenStatus(
      frozenAddresses: PromiseOrValue<string>[],
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<PopulatedTransaction>;
  };
}
